core/src/main/java/jenkins/model/Nodes.java

```diff
@@ -0,0 +1,254 @@
+            @Override
+            public boolean accept(File pathname) {
+                return pathname.isDirectory() && !existing.contains(pathname.getName());
+            }
+        })) {
+            Util.deleteRecursive(forDeletion);
+        }
+    }
+
+    /**
+     * Returns the named node.
+     *
+     * @param name the {@link Node#getNodeName()} of the node to retrieve.
+     * @return the {@link Node} or {@code null} if the node could not be found.
+     */
+    @CheckForNull
+    public Node getNode(String name) {
+        return nodes.get(name);
+    }
+
+    /**
+     * Loads the nodes from disk.
+     *
+     * @throws IOException if the nodes could not be deserialized.
+     */
+    public void load() throws IOException {
+        final File nodesDir = getNodesDir();
+        final File[] subdirs = nodesDir.listFiles(new FileFilter() {
+            public boolean accept(File child) {
+                return child.isDirectory();
+            }
+        });
+        final Map<String, Node> newNodes = new TreeMap<String, Node>();
+        if (subdirs != null) {
+            for (File subdir : subdirs) {
+                try {
+                    XmlFile xmlFile = new XmlFile(Jenkins.XSTREAM, new File(subdir, "config.xml"));
+                    if (xmlFile.exists()) {
+                        Node node = (Node) xmlFile.read();
+                        newNodes.put(node.getNodeName(), node);
+                    }
+                } catch (IOException e) {
+                    Logger.getLogger(ItemGroupMixIn.class.getName()).log(Level.WARNING, "could not load " + subdir, e);
+                }
+            }
+        }
+        Queue.withLock(new Runnable() {
+            @Override
+            public void run() {
+                for (Iterator<Map.Entry<String, Node>> i = nodes.entrySet().iterator(); i.hasNext(); ) {
+                    if (!(i.next().getValue() instanceof EphemeralNode)) {
+                        i.remove();
+                    }
+                }
+                nodes.putAll(newNodes);
+                jenkins.updateComputerList();
+                jenkins.trimLabels();
+            }
+        });
+    }
+
+    /**
+     * Returns the directory that the nodes are stored in.
+     *
+     * @return the directory that the nodes are stored in.
+     * @throws IOException
+     */
+    private File getNodesDir() throws IOException {
+        final File nodesDir = new File(jenkins.getRootDir(), "nodes");
+        if (!nodesDir.isDirectory() && !nodesDir.mkdirs()) {
+            throw new IOException(String.format("Could not mkdirs %s", nodesDir));
+        }
+        return nodesDir;
+    }
+
+    /**
+     * Returns {@code true} if and only if the list of nodes is stored in the legacy location.
+     *
+     * @return {@code true} if and only if the list of nodes is stored in the legacy location.
+     */
+    public boolean isLegacy() {
+        return !new File(jenkins.getRootDir(), "nodes").isDirectory();
```

Comment:
Would suggest a `@LocalData` test verifying compatibility with the legacy format.